slam-libraries/viam-cartographer/src/slam_service/slam_service.h

```diff
@@ -162,24 +145,88 @@ class SLAMServiceImpl final : public SLAMService::Service {
+                                    GetMapResponse *response);
+
+    // ProcessDataAndStartSavingMaps processes the data in the data directory
+    // that is newer than the provided data_cutoff_time
+    // and starts the process to save maps in parallel. In offline mode,
+    // all data in the directory is processed. In online mode, the most
+    // recently generated data is processed until a shutdown signal is
+    // received.
+    void ProcessDataAndStartSavingMaps(double data_cutoff_time);
+
+    // SetUpMapBuilder loads the lua file with default cartographer config
+    // parameters depending on the action mode.
+    void SetUpMapBuilder();
+
+    // GetLatestJpegMapString paints and returns the latest map as a jpeg string
+    // with or without the pose marker depending on the argument value.
+    std::string GetLatestJpegMapString(bool add_pose_marker);
+
+    // GetLatestSubmapSlices draws and returns the latest map as submap slices.
+    std::map<cartographer::mapping::SubmapId, ::cartographer::io::SubmapSlice>
+    GetLatestSubmapSlices();
+
+    // PaintMarker paints the latest global pose on the painted slices.
+    void PaintMarker(cartographer::io::PaintSubmapSlicesResult &painted_slices);
+
+    // ExtractPointCloudToString extracts the pointcloud from the map builder
+    // and saves it in a string. It returns a boolean that indicates whether
+    // or not the pointcloud string contains any points.
+    bool ExtractPointCloudToString(std::string &pointcloud_str);
+
+    // BackupLatestMap extracts and saves the latest map as a backup in
+    // the respective member variables.
+    void BackupLatestMap();
+
+    ActionMode action_mode = ActionMode::MAPPING;
+
     const std::string configuration_mapping_basename = "mapping_new_map.lua";
     const std::string configuration_localization_basename =
         "locating_in_map.lua";
     const std::string configuration_update_basename = "updating_a_map.lua";
+
     std::vector<std::string> file_list_offline;
     size_t current_file_offline = 0;
     std::string current_file_online;
 
-    std::atomic<bool> finished_processing_offline{false};
-    std::thread *thread_save_map_with_timestamp;
-
     std::mutex map_builder_mutex;
     mapping::MapBuilder map_builder;
 
-    std::mutex viam_response_mutex;
+    std::atomic<bool> finished_processing_offline{false};
+    std::thread *thread_save_map_with_timestamp;
+
+    std::mutex latest_global_pose_mutex;
     cartographer::transform::Rigid3d latest_global_pose =
         cartographer::transform::Rigid3d();
+
+    // --- The following variables are used exclusively to
+    // enable GetMap to send the most recent map out while
+    // cartographer works on creating an optimized map.
+    // The variables are only updated right before the
+    // optimization is started.
+    std::shared_mutex optimization_shared_mutex;
+
+    std::mutex latest_jpeg_map_str_mutex;
```

Comment:
[opt] I prefer combining `latest_global_pose_mutex`, `latest_jpeg_map_str_mutex`, and `latest_pointcloud_map_mutex` into a single `viam_response_mutex`, since these are used to control member variable accesses, which should all be fast. To me, that's less to think about, and it is more idiomatic to have a single mutex protecting member variable accesses.